packages/compartment-mapper/src/import-hook.js

```diff
@@ -399,3 +497,200 @@ export const makeImportHookMaker = (
+ * @param {string} baseLocation
+ * @param {MakeImportNowHookMakerOptions} options
+ * @returns {ImportNowHookMaker}
+ */
+export function makeImportNowHookMaker(
+  readPowers,
+  baseLocation,
+  {
+    sources = create(null),
+    compartmentDescriptors = create(null),
+    computeSha512 = undefined,
+    searchSuffixes = nodejsConventionSearchSuffixes,
+    sourceMapHook = undefined,
+    exitModuleImportNowHook,
+  },
+) {
+  // Set of specifiers for modules (scoped to compartment) whose parser is not
+  // using heuristics to determine imports.
+  /** @type {Map<string, Set<string>>} compartment name ->* module specifier */
+  const strictlyRequired = new Map();
+
+  /**
+   * @param {string} compartmentName
+   */
+  const strictlyRequiredForCompartment = compartmentName => {
+    let compartmentStrictlyRequired = strictlyRequired.get(compartmentName);
+    if (compartmentStrictlyRequired !== undefined) {
+      return compartmentStrictlyRequired;
+    }
+    compartmentStrictlyRequired = new Set();
+    strictlyRequired.set(compartmentName, compartmentStrictlyRequired);
+    return compartmentStrictlyRequired;
+  };
+
+  /**
+   * @type {ImportNowHookMaker}
+   */
+  const makeImportNowHook = ({
+    packageLocation,
+    packageName: _packageName,
+    entryCompartmentDescriptor,
+    parse,
+    compartments,
+  }) => {
+    const compartmentDescriptor = compartmentDescriptors[packageLocation] || {};
+
+    // this is necessary, because there's nothing that prevents someone from calling this function with the entry compartment.
+    assert(
+      compartmentDescriptor !== entryCompartmentDescriptor,
+      'Cannot create an ImportNowHook for the entry compartment',
+    );
+
+    // this is not strictly necessary, since the types should handle it.
+    assert(
+      parse[Symbol.toStringTag] !== 'AsyncFunction',
+      'async parsers are unsupported',
+    );
+
+    packageLocation = resolveLocation(packageLocation, baseLocation);
+    const packageSources = sources[packageLocation] || create(null);
+    sources[packageLocation] = packageSources;
+    const {
+      modules:
+        moduleDescriptors = /** @type {Record<string, ModuleDescriptor>} */ (
+          create(null)
+        ),
+    } = compartmentDescriptor;
+    compartmentDescriptor.modules = moduleDescriptors;
+
+    let { policy } = compartmentDescriptor;
+    policy = policy || create(null);
+
+    // associates modules with compartment descriptors based on policy
+    // which wouldn't otherwise be there
+    if ('packages' in policy && typeof policy.packages === 'object') {
+      for (const [packageName, packagePolicyItem] of entries(policy.packages)) {
+        if (
+          !(packageName in compartmentDescriptor.modules) &&
+          packageName in compartmentDescriptor.scopes &&
+          packagePolicyItem
+        ) {
+          compartmentDescriptor.modules[packageName] =
+            compartmentDescriptor.scopes[packageName];
+        }
+      }
+    }
```

Comment:
Feels like this would be better off as a function in policy.js called something like `addMissingRelationshipsToCompartmentModules` or something as meaningful but shorter.

Reply:
yeah, I agree

Reply:
though I don't know why it'd be in `policy.js`. I'm leaving it here for now

Reply:
please see changes from today